`get_start_shard` proposal

hwwhww opened the pull request:
### Issue
Implement `get_start_shard`

### How did I fix it
1. Add `current_epoch_start_shard` to `BeaconState`.
2. Implement `get_start_shard`
    1. Use `state.current_epoch_start_shard` as the current view, to plus or substract the committee count delta.
    2. `get_start_shard` and `get_committee_count_at_slot` results should be cached.
3. [**minimal config**] Change `INITIAL_ACTIVE_SHARDS` from `4` to `2` for more frequent crosslinks


TODO
- [x] add tests

terencechain commented on specs/phase1/beacon-chain.md:
```diff
@@ -530,33 +533,75 @@ def get_indexed_attestation(beacon_state: BeaconState, attestation: Attestation)
+    active_shard_count = get_active_shard_count(state)
+    if current_epoch_start_slot == slot:
+        return state.current_epoch_start_shard
+    elif current_epoch_start_slot > slot:
```
Should this be `slot > current_epoch_start_slot` ?  

`stop_slot` should be greater than `start_slot` for `get_committee_count_delta`

terencechain commented on specs/phase1/beacon-chain.md:
```diff
@@ -530,33 +533,75 @@ def get_indexed_attestation(beacon_state: BeaconState, attestation: Attestation)
+    else:
+        # Previous epoch
+        shard_delta = get_committee_count_delta(state, start_slot=slot, stop_slot=current_epoch_start_slot)
+        max_committees_per_epoch = MAX_COMMITTEES_PER_SLOT * SLOTS_PER_EPOCH
```
Optional 😅   In Prysm, I ended up doing 👇
```python
max_shards_per_epoch = MAX_COMMITTEES_PER_SLOT * SLOTS_PER_EPOCH * active_shard_count
return Shard(
            # Ensure positive
            (state.current_epoch_start_shard + max_shards_per_epoch - shard_delta)
            % active_shard_count
        )
```

hwwhww replied:
If we do so, the variable name `max_shards_per_epoch` will need to be changed. Any suggestion? :)

terencechain replied:
I see. On second though, I don't feel strongly about this one. We can leave it as it is

terencechain commented on specs/phase1/beacon-chain.md:
````diff
@@ -530,33 +533,75 @@ def get_indexed_attestation(beacon_state: BeaconState, attestation: Attestation)
+```python
+def get_committee_count_delta(state: BeaconState, start_slot: Slot, stop_slot: Slot) -> uint64:
+    """
+    Return the sum of committee counts between ``[start_slot, stop_slot)``.
````
`between` may not be accurate here. Since `[` means inclusive

hwwhww replied:
How about "between ``[start_slot, stop_slot)``" -> "in range ``[start_slot, stop_slot)``"?

djrtwo replied:
looks good

djrtwo commented on specs/phase1/beacon-chain.md:
````diff
@@ -993,9 +1038,19 @@ def process_epoch(state: BeaconState) -> None:
+#### Phase 1 final updates
+
+```python
+def process_phase_1_final_updates(state: BeaconState) -> None:
````
much cleanerrr :)

djrtwo commented on specs/phase1/beacon-chain.md:
```diff
@@ -993,9 +1038,19 @@ def process_epoch(state: BeaconState) -> None:
     process_light_client_committee_updates(state)
+
+    # Update current_epoch_start_shard
+    state.current_epoch_start_shard = get_start_shard(state, state.slot)
```
`process_epoch` happens before `state.slot` is incremented in the state transition so I think this is going to update the start shard to be the shard shard of `epoch_start_slot - 1` for the coming epoch.

Does it need to be the following?

```python
    state.current_epoch_start_shard = get_start_shard(state, state.slot + 1)
```

hwwhww replied:
Yes, you're right. 👍

hwwhww commented on specs/phase1/beacon-chain.md:
```diff
@@ -530,33 +533,75 @@ def get_indexed_attestation(beacon_state: BeaconState, attestation: Attestation)
+    """
+    Return the sum of committee counts between ``[start_slot, stop_slot)``.
+    """
+    committee_sum = 0
```
LGTM!
(I was debugging with `count` variable 😄)

djrtwo commented:
whoops, realized the `todo` : "Add tests" 


:)

hwwhww commented:
@djrtwo 

Added unit tests (not for test generator)!

Also, the minimal config's `INITIAL_ACTIVE_SHARDS` was set to `4`. With test suite default to 256 validators, the `committee_count_per_slot` would be only `2`. So I changed `INITIAL_ACTIVE_SHARDS` to `2` such that normally, we can crosslink all shards at every slot.

djrtwo commented on tests/core/pyspec/eth2spec/test/phase_1/unittests/test_get_start_shard.py:
```diff
@@ -0,0 +1,71 @@
+
+    current_epoch_start_slot = spec.compute_start_slot_at_epoch(spec.get_current_epoch(state))
+    expected_start_shard = (
+        state.current_epoch_start_shard +
+        spec.get_committee_count_delta(state, start_slot=current_epoch_start_slot, stop_slot=slot)
```
```suggestion
        state.current_epoch_start_shard
        + spec.get_committee_count_delta(state, start_slot=current_epoch_start_slot, stop_slot=slot)
```
